wincent/liferay-portal
# LPS-94647 Load dynamic-data-mapping-form-renderer CSS only when needed

The goal here is to stop the dynamic-data-mapping-form-renderer CSS from getting unconditionally loaded onto every single page (even when logged out, and no forms visible, for example). So, we get suppress the associated dynamic include and instead emit a stylesheet tag whenever we render a form. If there are multiple forms rendered, we will emit multiple tags, but only the first will be loaded by the browser and the others accessed directly as cached.

Thanks to [this Stack Overflow entry](https://stackoverflow.com/questions/6236097/is-link-not-rel-stylesheet-allowed-to-be-used-in-body) I was able to determine that HTML 5 [now allows `<link>` elements in the body, and all browsers accept them anyway](https://github.com/whatwg/html/commit/179983e9eb99efe417349a40ebb664bd11668ddd). The "property" attribute is added to allow the generated markup to pass validation.

This is a WIP for two or three reasons:

1. At the moment I have just commented out the unwanted unconditionally injected stylesheet tag in DDMFormRendererTopHeadDynamicInclude, just to show that this works; next step would be to completely remove the dynamic include class and all references to it.
2. As noted inline in a comment, for some reason I am getting the wrong context path in the form renderer ("/o/dynamic-data-mapping-form-web" instead of the desired "/o/dynamic-data-mapping-form-renderer") and I don't know why; I've overridden it with a janky assignment.
3. There is one other CSS file that needs the exact same treatment to be applied (the one from dynamic-data-mapping-web), but I wanted to validate this approach before doing the same thing over there.

Test plan:

- Hit portal logged out; see that unwanted CSS is no longer retrieved.
- Log in to control panel and see that it is still not retrieved.
- Go to site "Content & Data" -> "Forms" -> some form, and see CSS get loaded; form renders correctly.
- Likewise, click on "Preview Form" and see CSS get loaded.

One thing to note, the very first time you hit one of these pages after booting the portal you will see two requests hit the network for "main.css" instead of one. This is expected because the first is served directly by "AggregateFilter" unminified with a "Cache-Control: no-cache" header, while a minified version is written asynchonously to disk. The next request will get the minified version and be cached by the browser, by "CacheFilter" etc.
